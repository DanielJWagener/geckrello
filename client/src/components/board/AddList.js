import React from "react";
import { connect } from "react-redux";
import { addList } from "../../actions";

class AddList extends React.Component {
  state = {
    mode: "prompt",
    listTitle: "",
    borderColor: "#555",
    spans: 0
  };

  // Inialize element height on grid. Its grid height is based on CSS grid spans, so we fetch that value from the DOM and set the number of spans accordingly
  addListRef = React.createRef();

  setSpans = () => {
    const height = this.addListRef.current.clientHeight;

    const spans = Math.floor(height / 10);

    this.setState({ spans: spans });
  };

  componentDidMount() {
    this.setSpans();
  }

  // Mode toggle. This component has two modes: "prompt" and "input." The "prompt" mode is the initial mode, basically a button with a label like "add new list." When the user clicks that button, the mode changes to "input," and the user can input and submit a new list and title.
  toggleMode = () => {
    if (this.state.mode === "prompt") {
      this.setState({ mode: "input" });
    } else {
      this.setState({ mode: "prompt" });
    }
  };

  // Submit handlers
  onFormSubmit = event => {
    event.preventDefault();

    // Validate input
    if (!this.state.listTitle) {
      alert("Please enter a title.");
      return false;
    }

    // Generate a unique key and listId in the event of duplicate list titles
    const generateListId = inputId => {
      if (!this.props.lists.find(list => list.listId === inputId)) {
        return inputId;
      } else {
        inputId = `${inputId}-duplicate`;
        return generateListId(inputId);
      }
    };
    let listId = generateListId(this.state.listTitle);

    // Submit input
    this.props.addList(this.state.listTitle, listId);

    // Clear input
    this.setState({ listTitle: "", inputValid: false, borderColor: "#555" });

    // Change mode back to "prompt"
    this.toggleMode();
  };

  // Controlled input and validation to ensure unique list titles
  handleChange = e => {
    this.setState({
      listTitle: e.target.value
    });

    e.target.value
      ? this.setState({ borderColor: "#4fa644" })
      : this.setState({ borderColor: "red" });
  };

  render() {
    if (this.state.mode === "prompt") {
      return (
        <div ref={this.addListRef} className="addList addList__prompt">
          <button onClick={this.toggleMode} className="addList__prompt-button">
            <span className="addList__prompt-plus">+</span>
            <span className="addList__prompt-label"> Add new list</span>
          </button>
        </div>
      );
    } else {
      return (
        <div
          className="addList addList__input"
          ref={this.addListRef}
          style={{ gridRowEnd: `span ${this.state.spans}` }}
        >
          <div className="addList__form">
            <form onSubmit={this.onFormSubmit} className="form">
              <div className="form__group">
                <input
                  type="text"
                  className="form form__input form__input--validation"
                  value={this.state.listTitle}
                  onChange={this.handleChange}
                  style={{ borderBottomColor: this.state.borderColor }}
                  placeholder="New list title..."
                  autoFocus
                />
              </div>
              <div className="form__group">
                <input
                  type="submit"
                  className="form__submit"
                  value="Add List"
                />
              </div>
            </form>
          </div>
        </div>
      );
    }
  }
}

const mapStateToProps = state => {
  return { lists: state.lists };
};

export default connect(
  mapStateToProps,
  { addList }
)(AddList);
