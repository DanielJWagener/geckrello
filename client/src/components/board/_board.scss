.board {
  padding: 1rem;

  display: grid;
  grid-template-columns: repeat(auto-fill, minmax($list-width, 1fr));
  grid-gap: 0 1rem;
  grid-auto-rows: 1rem;
}

.addList {
  border-radius: 0.5rem;
  box-shadow: 0 1.5rem 4rem rgba(0, 0, 0, 0.3);
  height: 100px;
  z-index: 1;

  &__prompt {
    background-color: rgba(0, 0, 0, 0.6);

    &-button {
      height: 100%;
      width: 100%;
      color: white;
      background-color: transparent;
      border: none;
      cursor: pointer;
      font-size: 2rem;

      &:hover {
        color: #bbb;
      }
    }
  }

  &__input {
    background-color: $component-background;
  }

  &__form {
    width: 100%;
  }
}
