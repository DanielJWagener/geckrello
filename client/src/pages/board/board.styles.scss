@import "../../sass/abstracts/_variables.scss";

.board {
  height: calc(var(--vh, 1vh) * 100 - var(--navbar-height));

  &__lists-container {
    height: calc(
      var(--vh, 1vh) * 100 - var(--navbar-height) - #{$board-header-height} - 1rem
    );
    padding: 1rem 0;
    margin: 0 0.5rem;

    display: flex;
    overflow-x: auto;

    &::-webkit-scrollbar {
      width: 20px;
      margin: 0 1rem 3rem 1rem;
      padding: 0 1rem;
    }

    &::-webkit-scrollbar-track {
      -webkit-border-radius: 5px;
      border-radius: 5px;
      background: #aaa;
    }

    &::-webkit-scrollbar-thumb {
      -webkit-border-radius: 5px;
      border-radius: 5px;
      background: #fff;
      margin: 0 1rem;

      &:hover {
        cursor: pointer;
      }

      &:active {
        background: #eee;
      }
    }
  }
}
