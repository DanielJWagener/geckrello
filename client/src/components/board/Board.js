import AddList from "./AddList";
import List from "./list/List";
import React from "react";
import { DndProvider } from "react-dnd";
import HTML5Backend from "react-dnd-html5-backend";
import { connect } from "react-redux";

class Board extends React.Component {
  // Iterate over every list in state, return the non-archived ones, and make and array of List components out of them
  listsArray = () =>
    this.props.lists
      .filter(list => !list.archived)
      .map(list => (
        <List
          key={list.listId}
          listTitle={list.title}
          listId={list.listId}
          cards={[]}
        />
      ));

  render() {
    return (
      <DndProvider backend={HTML5Backend}>
        <div className="board" id="board">
          {this.listsArray()}
          <AddList />
        </div>
      </DndProvider>
    );
  }
}

const mapStateToProps = state => {
  return { lists: state.lists };
};

export default connect(mapStateToProps)(Board);
