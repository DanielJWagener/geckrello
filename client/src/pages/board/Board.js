import React from "react";
import { DndProvider } from "react-dnd";
import HTML5Backend from "react-dnd-html5-backend";
import { connect } from "react-redux";

import AddList from "../../components/AddList";
import List from "../../components/List/list.component";
import MenuBar from "../../components/menu-bar/MenuBar";
import { unloadBoard, fetchBoardData } from "../../actions";
import colorThemes from "../../utilities/colorThemes";

class Board extends React.Component {
  componentDidMount() {
    this.props.fetchBoardData(this.props.match.params.id);
  }

  componentDidUpdate() {
    let root = document.documentElement;

    if (this.props.board) {
      const theme = colorThemes[this.props.board.background];

      root.style.setProperty("--color-primary", theme.base);
      root.style.setProperty("--color-primary-lightest", theme.lightest);
      root.style.setProperty("--color-primary-lighter", theme.lighter);
      root.style.setProperty("--color-primary-light", theme.light);
      root.style.setProperty("--color-primary-dark", theme.dark);
      root.style.setProperty("--color-primary-darker", theme.darker);
      root.style.setProperty("--color-primary-darkest", theme.darkest);
    }
  }

  componentWillUnmount() {
    this.props.unloadBoard();
  }

  // Iterate over every list in state, return the non-archived ones, and make and array of List components out of them
  listsArray = () =>
    this.props.lists
      .filter(list => !list.archived)
      .map(list => (
        <List
          key={list._id}
          listTitle={list.title}
          listId={list._id}
          cards={[]}
          archived={list.archived}
        />
      ));

  render() {
    if (this.props.board) {
      return (
        <div>
          <MenuBar board={this.props.board} />
          <DndProvider backend={HTML5Backend}>
            <div className="board" id="board">
              {this.listsArray()}
              <AddList boardId={this.props.match.params.id} />
            </div>
          </DndProvider>
        </div>
      );
    } else {
      return <h1>Loading...</h1>;
    }
  }
}

const mapStateToProps = ({ board, lists, cards }) => {
  return { board, lists, cards };
};

export default connect(mapStateToProps, {
  unloadBoard,
  fetchBoardData
})(Board);
