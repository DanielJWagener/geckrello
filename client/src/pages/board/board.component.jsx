import React from "react";
import { DndProvider } from "react-dnd";
import HTML5Backend from "react-dnd-html5-backend";
import { connect } from "react-redux";

import AddList from "../../components/AddList/add-list.component";
import List from "../../components/List/list.component";
import BoardHeader from "../../components/BoardHeader/board-header.component";
import { unloadBoard, fetchBoardData } from "../../redux/boards/boards.actions";
import colorThemes from "../../utilities/colorThemes";
import updateBackgroundColor from "../../utilities/updateBackgroundColor";
import Spinner from "../../components/Spinner/spinner.component";
import { selectUnarchivedLists } from "../../redux/lists/lists.selectors";

import "./board.styles.scss";

export class Board extends React.Component {
  componentDidMount() {
    this.props.board &&
      updateBackgroundColor(colorThemes[this.props.board.background]);

    this.props.fetchBoardData(this.props.match.params.id);
  }

  componentDidUpdate() {
    if (this.props.board) {
      const theme = colorThemes[this.props.board.background];

      updateBackgroundColor(theme);
    }
  }

  componentWillUnmount() {
    this.props.unloadBoard();
  }

  // Iterate over every list in state, return the non-archived ones, and make an array of List components out of them
  listsArray = () =>
    this.props.lists.map(list => (
      <List
        key={list._id}
        listTitle={list.title}
        listId={list._id}
        cards={[]}
        archived={list.archived}
      />
    ));

  render() {
    if (this.props.board && !this.props.board.isPending) {
      return (
        <>
          <DndProvider backend={HTML5Backend}>
            <div className="board" id="board">
              <BoardHeader board={this.props.board} />
              <div className="board__lists-container">
                {this.listsArray()}
                <AddList boardId={this.props.match.params.id} />
              </div>
            </div>
          </DndProvider>
        </>
      );
    } else {
      return <Spinner />;
    }
  }
}

const mapStateToProps = state => {
  return {
    board: state.board,
    lists: selectUnarchivedLists(state)
  };
};

export default connect(mapStateToProps, {
  unloadBoard,
  fetchBoardData
})(Board);
