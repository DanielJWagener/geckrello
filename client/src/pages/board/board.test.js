import { shallow } from "enzyme";
import React from "react";

import { Board as BoardPure } from "./board.component";

it("renders Board component", () => {
  const mockProps = {
    match: { params: { id: 1 } },
    fetchBoardData: () => {},
  };

  expect(shallow(<BoardPure {...mockProps} />).debug()).toMatchSnapshot();
});
