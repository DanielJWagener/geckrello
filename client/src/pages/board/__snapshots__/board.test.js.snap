// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Board component 1`] = `"<Spinner />"`;
